### DISCUSSION TITLE
Using neomodel's data models in place of pydantic

### DISCUSSION THREAD

**Discussion by dvjn** [fastapi/fastapi#7179] [Questions]
I am making a `fastapi` server with `neo4j` database, and am using the `neomodel` driver for connecting to the database.
I will have to create 2 seperate models, first in neomodel, second in pydantic (for request body validation and stuff). Is there a way to use the neomodel models in all the places where pydantic models are used in fastapi, if not could anyone point me in a direction to make the development experience as frictionless as possible.

Also, I am open to making a PR (if this is a feature that can be implemented), I'll just need a little guidance because I don't know the fastapi codebase.

BTW, thanks for the amazing library!

**Comment by Kludex** [comment]
I don't think there's space for that on the FastAPI code source. I see it as analogous of what we do when we use SQLAlchemy: create both Pydantic and SQLAlchemy models. 

What I've checked is that there's a package called [django-neomodel](https://github.com/neo4j-contrib/django-neomodel) (last commit 2 years ago) that could give you ideas on a possible package called `fastapi-neomodel`. Similar to what @tiangolo is doing for SQLAlchemy + Pydantic.

**Comment by dvjn** [comment]
Thank you for pointing me in the right direction!

@tiangolo's `pydantic-sqlalchemy` is the inspiration that I was looking for, the codebase looks relatively simple, to use for making a `pydantic-neomodel` library, but then it seems like I will lose so many features in pydantic like constraints, because there is limited support for most of them in neomodel (afaik).

Maybe asking for pydantic models support in neomodel is a better approach. Your thoughts?

**Comment by Kludex** [comment]
To be honest, `pydantic-sqlalchemy` was not the one I was talking about. I was talking about [this](https://twitter.com/tiangolo/status/1287679028036599809). It's not available yet, and I don't know if it will, but the idea I was referring is that one.

> but then it seems like I will lose so many features in pydantic like constraints, because there is limited support for most of them in neomodel (afaik).

Well, if you want an interface between those two, it seems fair to limit one based on the other.

> Maybe asking for pydantic models support in neomodel is a better approach. Your thoughts?

I don't see why not. :+1:

**Comment by dvjn** [comment]
Damn, **[this](https://twitter.com/tiangolo/status/1287679028036599809)** is nice. I'd love to be able to do that. 
Hope this turns into a feature soon.

Anyways, thanks for your time and help @Kludex ♥
And, thanks for the library @tiangolo 🙌

**Comment by NFEL** [comment]
Hey Guys, i really to integrate a feature like this as a plug-in feature to pydantic.
Much like [Odmantic](https://art049.github.io/odmantic/), field mappings can be done to integrate default python + some extra fields around [neomodel-properties](https://neomodel.readthedocs.io/en/latest/properties.html#property-types).
I like to work on a project like this ....

**Comment by Murat-U-Saglam** [comment]
Did anyone end up doing something?

**Reply by NFEL** [reply]
Well, i did but it end up being a mess and domain specific for work I was doing.
The issue is sql has a very straight forward approach when it comes to data. but Neo4J sql-like queries are much harder to generalize without much though.

**Comment by Murat-U-Saglam** [comment]
https://github.com/ontolocy/neontology

**Reply by hardchor** [reply]
I'm personally using this one, but it's extremely limited compared to a full OGM like neomodel. I think the author is explicitly asking for a way to integrate pydantic models with neomodel (which makes so much sense in my book)

**Reply by Pk13055** [reply]
@hardchor Thanks for your comment, I was seriously evaluating shifting to this, given that I'm running into schema defining issues when connecting the backend, but if it is limited I think better to hold off for now. Have you maybe tried [pydantic-neomodel-dict](https://github.com/HardMax71/pydantic-neomodel-dict) for simpler cross-integration?

**Comment by Pk13055** [comment]
For now, if your use case is simple, ie you just want to render the responses directly, you can use the following function to wrap your query outputs using `neomodel` OGM results:
```python
from neomodel.async_.node import NodeMeta
from neomodel.properties import Property


async def encode_payload(payload: NodeMeta | list[NodeMeta]) -> dict | list[dict]:
    """Encode a payload for Neo4j"""
    # get properties (except relationships)
    print(type(payload))
    response = None
    _repr = payload[0] if isinstance(payload, list) else payload
    props = set(
        {k for k, v in _repr.defined_properties().items() if isinstance(v, Property)}
    )
    if isinstance(payload, list):
        response = [
            {k: v for k, v in item.__dict__.items() if k in props} for item in payload
        ]
    else:
        response = {k: v for k, v in payload.__dict__.items() if k in props}
    return response


# Simple example:
from neomodel import AsyncStructuredNode, StringProperty, AsyncRelationshipTo, AsyncZeroOrMore
class Period(AsyncStructuredNode):
    """Period node for temporal organization"""

    period = StringProperty(
        required=True,
        unique_index=True,
        choices={"Q1": "Q1", "Q2": "Q2", "Q3": "Q3", "Q4": "Q4", "FY": "FY"},
    )

    # relationships
    metrics = AsyncRelationshipTo(
        "StatementMetrics", "HAS_METRICS", cardinality=AsyncZeroOrMore
    )

async def test():
    return await encode_payload(await Period.nodes.filter(period__in=['Q1, 'Q2']).all())
```
